File: py_drill/Problems/PQ04_KClosestPointsToOrigin.py
# REF : https://leetcode.com/problems/k-closest-points-to-origin/
# NOTES:
#       Two ways to do -
#       1. create maxHeap of size k and after popping n-k elements,
#          last k remaining elements in the maxHeap will be answers. O(k + (n-k)logk)
#       2. create minHeap with all elements from the list and keep popping till k. That'll be ans.
#          O(n + klogn), n -> heapify AND klogn for popping k times. BETTER
#       Heap elements will be (x*x + y*y, x, y). sorting will be based on Euclidean distance.

import heapq

class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:

        # T: O(n + klogn), n -> for heapify AND klogn for popping k elements from minHeap
        # S: O(n)
        minHeap = []
        for x, y in points:
            dist = x*x + y*y                # x**2 + y**2
            minHeap.append([dist, x, y])

        heapq.heapify(minHeap)
        # pop k elements
        result = []
        while k > 0:
            dist, x, y = heapq.heappop(minHeap)
            result.append([x, y])
            k -= 1

        return result


def main():
    points = [[3,3],[5,-1],[-2,4]]
    k = 2
    print("output : {}".format(Solution().kClosest(points, k)))

if __name__ == "__main__":
    main()

File: cpp_drill/Problems/PQ04_KClosestPointsToOrigin.cpp
#include <iostream>
#include <queue>
using namespace std;

// REF : https://leetcode.com/problems/k-closest-points-to-origin/
// NOTES :
//      Min Heap would require pushing all the elements and popping top k elements.
//      Max Heap of size k & popping top elements if size reaches k.
//      remaining elements in the Max Heap will be answer.

typedef pair<int, pair<int, int>> mypair;

// T: O(n + klogn) n -> pushing all elements & k times popping logn
// S: O(n)
// Using Min Heap
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<mypair, vector<mypair>, greater<mypair>> minHeap;
        for (auto &i : points) {
            int dist = i[0]*i[0] + i[1]*i[1];
            minHeap.push({dist, {i[0], i[1]}});
        }

        vector<vector<int>> result;
        while (k > 0) {
            mypair element = minHeap.top();
            minHeap.pop();
            result.push_back({element.second.first, element.second.second});
            k--;
        }
        return result;
    }
};

// T: O(nlogk)
// S: O(k)      Max Heap will be maintaining k elements only
// Using Max Heap
class Solution1 {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        priority_queue<mypair> maxHeap;
        for (auto &i : points) {
            int dist = i[0]*i[0] + i[1]*i[1];
            maxHeap.push({dist, {i[0], i[1]}});
            if (maxHeap.size() > k)
                maxHeap.pop();
        }

        vector<vector<int>> result;
        while (!maxHeap.empty()) {
            mypair element = maxHeap.top();
            maxHeap.pop();
            result.push_back({element.second.first, element.second.second});
        }
        return result;
    }
};

int main() {
    vector<vector<int>> points = {{3,2}, {7,7}, {-9,9}, {4,-6}, {-3,-6}};
    int k = 4;
    vector<vector<int>> result;

    result = Solution1().kClosest(points, k);
    for (auto &i : result) {
        for (auto &j : i) {
            cout << j << " ";
        }
        cout << endl;
    }
}